### PR TITLE
Mobx refactor for bot-court and remove EventCardsGridContainer(blocked)

### PR DESCRIPTION
**Pull request by amazingandyyy**

### CODE REVIEW

**Comment by amazingandyyy** [comment]
refer to #682

File: src/stores/BotCourtStore.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        this.list = [...this.list, ...nextFewEvents];
+        this.loadingMore = false;
+      });
+    }
+  }
+
+  async fetch(limit = this.limit, skip = this.skip) {
+    const orderBy = { field: 'endTime', direction: this.app.sortBy };
+    const excludeResultSetterQAddress = this.app.wallet.addresses.map(({ address }) => address);
+    const filters = [
```
Comment (dwalintukan) [inline src/stores/BotCourtStore.js]: move all these vars inside `if (this.hasMore) {`

File: src/stores/BotCourtStore.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      { token: Token.Qtum,
+        status: OracleStatus.WaitResult,
+        excludeResultSetterQAddress,
+      },
+    ];
+    let oracles = [];
+    if (this.hasMore) {
+      oracles = await queryAllOracles(filters, orderBy, limit, skip);
+      oracles = _.uniqBy(oracles, 'txid').map((oracle) => new Oracle(oracle, this.app));
+      if (oracles.length < limit) this.hasMore = false;
```
Comment (dwalintukan) [inline src/stores/BotCourtStore.js]: `if (oracles.length < this.skip)`

this check should be dependent on the current "page" it is on. eg. if its on page 2, it should have 32 items. so you should be checking if the 2nd page pull have less than 32.

File: src/stores/BotCourtStore.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    const botCourt = _.orderBy(oracles, ['endTime'], this.app.sortBy.toLowerCase());
```
Comment (dwalintukan) [inline src/stores/BotCourtStore.js]: can you add a separate `sortBy` for this component instead of having a global one? later there will be many different sorting options and some pages might not have the same sorting options.

File: src/components/EventCardsGridContainer/index.js [hunk 4]
```diff
@@ -1,136 +0,0 @@
...
-  }
-
-  componentWillReceiveProps(nextProps) {
-    const { eventStatusIndex, sortBy, syncBlockNum, walletAddresses, txReturn } = nextProps;
-
-    if (eventStatusIndex !== this.props.eventStatusIndex
-      || sortBy !== this.props.sortBy
-      || syncBlockNum !== this.props.syncBlockNum
-      || walletAddresses !== this.props.walletAddresses
-      || txReturn !== this.props.txReturn) {
```
Comment (dwalintukan) [inline src/components/EventCardsGridContainer/index.js]: you are missing this logic in your BotCourt component:
```
|| syncBlockNum !== this.props.syncBlockNum
|| walletAddresses !== this.props.walletAddresses
|| txReturn !== this.props.txReturn) {
```

These are all important and they cause the page to update on:
1. a new block
2. when the wallet addresses change
3. when a tx is executed.

Comment (amazingandyyy) [reply src/components/EventCardsGridContainer/index.js]: wow, hm actually all 4 components who used `EventCardsGridContainer` are all missing now.... jesus. I will discuss this with alex about how to setup the MobX store structure for these things...

**Comment by amazingandyyy** [comment]
will rebase soon as well

File: src/stores/BotCourtStore.js (hunk 2 above)
Comment (dwalintukan) [reply src/stores/BotCourtStore.js]: Gotcha my bad. Thought it was the full list.

File: src/stores/BotCourtStore.js [hunk 5]
```diff
@@ -4,37 +4,47 @@ import { Token, OracleStatus, AppLocation } from '../constants';
...
+};
 
 export default class BotCourtStore {
-  @observable loading = true
-  @observable loadingMore = false
-  @observable list = []
-  @observable hasMore = true
-  @observable skip = 0
-  limit = 16
+  @observable loaded = INIT.loaded
```
Comment (dwalintukan) [inline src/stores/BotCourtStore.js]: is `loaded` ever changed?

File: src/stores/BotCourtStore.js [hunk 6]
```diff
@@ -4,37 +4,47 @@ import { Token, OracleStatus, AppLocation } from '../constants';
...
-    if (limit === this.limit) {
-      this.skip = 0;
-    }
-    this.hasMore = true;
+    this.reset(); // reset to initial state
     this.app.ui.location = AppLocation.botCourt;
     this.list = await this.fetch(limit);
     runInAction(() => {
-      this.skip += limit;
       this.loading = false;
```
Comment (dwalintukan) [inline src/stores/BotCourtStore.js]: changed to `this.loaded = true`?

File: src/components/EventCardsGridContainer/index.js (hunk 4 above)
Comment (dwalintukan) [reply src/components/EventCardsGridContainer/index.js]: Hmmm so its missing all this refreshing logic then... So for the time being, you will have to use the same logic with `componentWillReceiveProps` to make this work with the ones that aren't converted to MobX.

We already have `addresses` in the `WalletStore` so you will have to make a reaction in `BotCourtStore`:
```
reaction(
  () => this.app.wallet.addresses
  () => this.init() // refresh all the items
);
```

I have a PR for the `GlobalStore` that has the `syncBlockNum` so you can add a reaction for that as well when the PR gets merged.

And `txReturn` isnt converted to MobX yet so you will have to use the same logic as the deleted logic for now until that gets converted.

File: src/stores/BotCourtStore.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    );
+  }
+
+  @action
+  init = async (limit = this.limit) => {
+    this.reset(); // reset to initial state
+    this.app.ui.location = AppLocation.botCourt;
+    this.list = await this.fetch(limit);
+    runInAction(() => {
+      this.loading = false;
```
Comment (dwalintukan) [inline src/stores/BotCourtStore.js]: supposed to be `this.loaded = true`?

Comment (amazingandyyy) [reply src/stores/BotCourtStore.js]: updated, here is where we change loaded, and hm sadly eslint didn't catch this kind, will be careful next time.

File: src/stores/BotCourtStore.js (hunk 5 above)
Comment (amazingandyyy) [reply src/stores/BotCourtStore.js]: updted

Comment (amazingandyyy) [reply src/stores/BotCourtStore.js]: updated

File: src/components/EventCardsGridContainer/index.js (hunk 4 above)
Comment (amazingandyyy) [reply src/components/EventCardsGridContainer/index.js]: oh, that's awesome, ok I will update them all and add reaction to every relevant components

**Comment by amazingandyyy** [comment]
alex has some concerns about merging `UiStore` and `pubSubStore` to a `GlobalStore`,  I will come back to this PR and update the `txReturn` later. I am thinking move txRetrun into `GlobalStore` or `TransactionStore`.

File: src/components/EventCardsGridContainer/index.js (hunk 4 above)
Comment (amazingandyyy) [reply src/components/EventCardsGridContainer/index.js]: the whole idea of `txReturn` as a global variable doesn't make sense, discussed with alex and henry this afternoon. we decide to have a local variable like transactionFinished? to trigger an init.

Comment (dwalintukan) [reply src/components/EventCardsGridContainer/index.js]: sounds good

File: src/scenes/BotCourt/styles.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+const styles = () => ({
+  row: {
+    display: 'flex',
+    justifyContent: 'center',
+    alignItems: 'center',
+    width: '100%',
+    margin: 20,
```
[the remark this answers is not in the document]
Comment (dwalintukan) [reply src/scenes/BotCourt/styles.js]: generally, try to use the predefined theme values first. unless its a one-off case where its not used anywhere else.

File: src/scenes/Event/components/EventTxHistory/TxRow.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+import TransactionHistoryID from '../../../../components/TransactionHistoryAddressAndID/id';
+import TransactionHistoryAddress from '../../../../components/TransactionHistoryAddressAndID/address';
+import { getShortLocalDateTimeString } from '../../../../helpers/utility';
+import { localizeInvalidOption } from '../../../../helpers/localizeInvalidOption';
+import { getTxTypeString } from '../../../../helpers/stringUtil';
+import { TransactionType } from '../../../../constants';
+
+@injectIntl
+@withStyles(styles, { withTheme: true })
+export default class TxRow extends Component {
```
Comment (dwalintukan) [inline src/scenes/Event/components/EventTxHistory/TxRow.js]: this looks to be changes from another PR. i think the rebase went wrong.

File: src/components/WalletUnlockDialog/index.js [hunk 10]
```diff
@@ -27,41 +25,32 @@ const messages = defineMessages({
...
-    walletUnlockDialogVisibility: PropTypes.bool.isRequired,
-    toggleWalletUnlockDialog: PropTypes.func.isRequired,
-    unlockWallet: PropTypes.func.isRequired,
     intl: intlShape.isRequired, // eslint-disable-line react/no-typos
   };
 
-  state = {
-    passphrase: '',
-    unlockMinutes: Config.defaults.unlockWalletMins,
+  unlock = () => {
```
Comment (dwalintukan) [inline src/components/WalletUnlockDialog/index.js]: these are changes from a PR I already did.

File: src/stores/AllEventsStore.js [hunk 11]
```diff
@@ -29,12 +29,25 @@ export default class {
...
         if (this.app.ui.location === AppLocation.allEvents) {
+          this.init();
+        }
+      }
+    );
+    reaction(
+      () => this.app.sortBy,
+      () => {
+        if (this.app.ui.location === AppLocation.allEvents) {
+<<<<<<< HEAD
```
Comment (dwalintukan) [inline src/stores/AllEventsStore.js]: fix merge conflict

File: src/stores/BotCourtStore.js [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  @observable loadingMore = INIT.loadingMore
+  @observable list = INIT.list
+  @observable hasMore = INIT.hasMore
+  @observable skip = INIT.skip
+  limit = INIT.limit
+
+  constructor(app) {
+    this.app = app;
+    reaction(
+      () => this.app.sortBy,
```
Comment (dwalintukan) [inline src/stores/BotCourtStore.js]: need to watch `blockNum` and `addresses` here as well.

File: src/stores/BotCourtStore.js [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+      ];
+      let oracles = [];
+      oracles = await queryAllOracles(filters, orderBy, limit, skip);
+      oracles = _.uniqBy(oracles, 'txid').map((oracle) => new Oracle(oracle, this.app));
+      return _.orderBy(oracles, ['endTime'], this.app.sortBy.toLowerCase());
+    }
+    return INIT.list;
+  }
+
+  reset = () => {
```
Comment (dwalintukan) [inline src/stores/BotCourtStore.js]: Alex taught a cool trick: `reset = () => Object.assign(this, INIT)` so you dont need to manually set them.

File: src/stores/activitiesStores/Finalize.js [hunk 14]
```diff
@@ -23,6 +23,16 @@ export default class {
 
   constructor(app) {
     this.app = app;
+    const { wallet: { addresses } } = this.app;
+    const { syncBlockNum } = this.global;
+    reaction(
+      () => addresses + syncBlockNum,
```
Comment (dwalintukan) [inline src/stores/activitiesStores/Finalize.js]: 👍

File: src/components/EventCardsGridContainer/index.js (hunk 4 above)
Comment (dwalintukan) [reply src/components/EventCardsGridContainer/index.js]: `QtumPredicition` is the only place that should have the `txReturn` mobx reaction. This is when the user creates a new event, the page should refresh based on the `txReturn` of the CreateEvent call.

**Comment by amazingandyyy** [comment]
@dwalintukan  I add a `refeshing` to AppStore. pls review and I will ship it soon. thank you.

File: src/components/EventCardsGridContainer/index.js (hunk 4 above)
Comment (amazingandyyy) [reply src/components/EventCardsGridContainer/index.js]: use freshing variable for it.

File: src/stores/RefreshingStore.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
+import { observable, action } from 'mobx';
+
+const INIT_VALUES = {
+  status: false,
+  type: '',
+};
+
+export default class {
```
Comment (dwalintukan) [inline src/stores/RefreshingStore.js]: I assume this is the replacement for the `txReturn` refreshing logic?
